### PR TITLE
Add `retryWhen`

### PR DESCRIPTION
**Pull request by inamiy**
This is an attempt to implement [`retryWhen`](http://reactivex.io/documentation/operators/retry.html) in ReactiveSwift.
I wrote simple version & Rx original version here:

- (1)  8754889: Simpler version with type: 
    - `(FlattenStrategy, (Error) -> SignalProducer<(), Error>) -> SignalProducer<Value, Error>`
- (2) 2f829f8: Rx original version with type:
    - `(FlattenStrategy, (Signal<Error, NoError>) -> Signal<(), Error>) -> SignalProducer<Value, Error>`

### Use case: Delay retry after error

```swift
// Using (1) Simple version
producer.retry(.merge) { error in
    return timer(3, on: mainScheduler).take(1).map { _ in () } // retry after 3 sec
}

// Using (2) Rx original version
producer.retry(.merge) { errors in
    return errors.flatMap { 
        return timer(3, on: mainScheduler).take(1).map { _ in () } // retry after 3 sec
    }.take(5)  // up to 5 times
}
```

(1) is obviously simpler and has a similar type as `flatMap` (which I think is clean), but it does not have error counting logic so (2) might be a better option.

But to implement (2), I had to change `NSLock` to `NSRecursiveLock` in `Signal.swift`.
(BTW I personally think using recursive-lock is a better choice in general)

### CODE REVIEW

**Comment by inamiy** [comment]
```
FlattenStrategy_concurrent__should_asynchronously_merge_up_to_the_stated_limit__buffer_any_subsequent_producers_and_dequeue_them_in_the_submission_order, expected to equal <[0, 1, 2, 3]>, got <[0]>
```

CI error seems unrelated to this PR.

**Comment by mdiep** [comment]
A few notes:

1. As an additive API, this doesn't need to hold up a 2.0 release.

2. I think the simpler `(Error) -> SignalProducer<(), Error>` variant is preferable.

3. I don't think this needs to take a flatten strategy. It seems like the returned signal should have a `take(first: 1)` that prevents multiple retries from the same error.

4. I don't think a recursive lock is going to be acceptable.

Do you have a particular use case in mind for this? Or are you suggesting it because you find it interesting?

**Comment by inamiy** [comment]
@mdiep 
Use cases will be:
1. Delaying retry (as I just showed example on top comment)
2. Conditional retry based on `error` (e.g. retry only when error has status code XXX...)

> I don't think this needs to take a flatten strategy

Sounds simpler! I will just use `.merge` then 😉 

If we go with (1), I also need error counting to stop retrying forever, so below code might be better: 

```swift
public func retry(upTo count: Int, when: @escaping (Error) -> SignalProducer<(), Error>) -> SignalProducer<Value, Error> {
	precondition(count >= 0)

	if count == 0 {
		return producer
	} else {
		return flatMapError { error in
			return when(error).flatMap(.merge) {
				return self.producer.retry(upTo: count - 1, when: when)
			}
		}
	}
}
```

**Comment by mdiep** [comment]
I wonder if we'd be better adding variants that cover those two cases. Those seem like the most common uses.

The first case would actually be really well suited to have some sort of timing function—an enum that covered both a constant time and an exponential backoff with some sort of upper limit.

**Comment by inamiy** [comment]
@mdiep 
Adding variants sounds friendly, but for exponential backoff, I guess it will require additional state inside which will make (1)'s implementation more complicated.

On the other hand, if we go with (2), we can use `scan` to count errors and handle exponential delay calculation (and probably many more).

So... Though I wasn't aware of (2)'s powerfulness on the first place, I'm leaning toward (2) now 🤔

**Comment by inamiy** [comment]
Updated the code to track current master branch.

Signature became simpler as follows, but still, `PthreadLock(recursive: true)` is required to avoid the crash in test cases.

```swift
func retry(when: ((Signal<Error, NoError>) -> Signal<(), Error>) -> SignalProducer<Value, Error>
```

File: Sources/Signal.swift [hunk 1]
```diff
@@ -67,7 +67,7 @@ public final class Signal<Value, Error: Swift.Error> {
 			state = .alive(Bag(), hasDeinitialized: false)
 
 			stateLock = Lock.make()
-			sendLock = Lock.make()
+			sendLock = Lock.PthreadLock(recursive: true)
```
Comment (chuganzy) [inline Sources/Signal.swift]: Agree with this change. Ref: https://github.com/ReactiveCocoa/ReactiveSwift/issues/74#issuecomment-472686906

**Comment by andersio** [comment]
We should look at improving options for retrying, though use of recursive locks is a non-goal.

Closing due to inactivity.

**Pull request closed without merging**